File: app/src/main/java/com/apptest/adapter/UsersAdapter.java
package com.apptest.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.apptest.R;
import com.apptest.models.User;
import com.apptest.viewholder.UserViewHolder;

import java.util.ArrayList;

/**
 * Created by davidpayel on 04/02/2017.
 */

public class UsersAdapter extends RecyclerView.Adapter<UserViewHolder> {
    private ArrayList<User> users;
    private Context context;
    private OnItemClickListener mListener;

    public void onItemClickListener(OnItemClickListener mListener) {
        this.mListener = mListener;
    }

    public UsersAdapter() {
        this.users = new ArrayList<>();
    }

    public UsersAdapter(Context context, ArrayList<User> users, OnItemClickListener listener) {
        this.context = context;
        this.users = users;
        this.mListener = listener;
    }

    public void setUsers(ArrayList<User> users) {
        this.users = users;
    }

    public interface OnItemClickListener {
        void onItemClick(User user, View v);
    }

    @Override
    public UserViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.user_adapter, parent, false);
        return new UserViewHolder(context, view);
    }

    @Override
    public void onBindViewHolder(UserViewHolder holder, int position) {
        //User user = users.get(position);
        //holder.bind(user, position, mListener);
    }

    @Override
    public int getItemCount() {
        return users.size();
    }
}

File: app/src/main/java/com/apptest/Users/adapters/UsersAdapter.kt
package com.apptest.Users.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.apptest.R
import com.apptest.Users.models.User

import java.util.ArrayList

/**
 * Created by davidpayel on 04/02/2017.
 */

class UsersAdapter : RecyclerView.Adapter<UserViewHolder> {
    private var users: ArrayList<User>? = null
    private var context: Context? = null
    private var mListener: OnItemClickListener? = null

    constructor(context: Context, listener: OnItemClickListener) {
        this.context = context
        this.mListener = listener
    }

    fun setUsers(users: ArrayList<User>) {
        this.users = users
    }

    interface OnItemClickListener {
        fun onItemClick(user: User, v: View)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.user_adapter, parent, false)
        return UserViewHolder(context!!, view)
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        val user = users!![position]
        holder.bind(user, position, mListener)
    }

    override fun getItemCount(): Int {
        return users!!.size
    }
}